gnovm/pkg/gnolang/op_expressions.go

```diff
@@ -230,16 +255,22 @@ func (m *Machine) doOpTypeAssert1() {
 		} else if nt, ok := baseOf(t).(*NativeType); ok {
 			// t is Go interface.
 			// assert that x implements type.
-			impl := false
+			nonConcrete := "non-concrete "
+			var impl bool
 			if nxt, ok := xt.(*NativeType); ok {
-				impl = nxt.Type.Implements(nt.Type)
-			} else {
-				impl = false
+				// If the underlying native type is reflect.Interface kind, then this has no
+				// concrete value and should fail.
+				if nxt.Type.Kind() != reflect.Interface {
+					impl = nxt.Type.Implements(nt.Type)
+					nonConcrete = ""
+				}
```

Comment:
So, I don't think this assertion is fully correct and I would revert it to the previous one. Although... it's hard to think of any situation where this matters.

The situations where a `Type.Kind()` is an interface, but the value is actually concrete, are simply when the interface is "nested" within a type: https://go.dev/play/p/eKwnf3ncyti

So, it seems to me, that possibly one of the few places where this _could_ happen, is we tried to get `flag.Flag.Value`? I'm not entirely sure even then. We'd need an `interface{}` wrapped within some other kind of compound value

Reply:
I don't quite understand all of that, but I agree that it is hard to think of situations where this matters. I think it matters for the sake of being consistent with the go spec.

The file test type37a_native.gno exemplifies how, in some cases, the type assertion should fail with native types, even if it is a typed nil value. For example, this should fail because there is no concrete value:
```go
var v interface{}
var r io.Reader
v = r
_ = r.(io.Reader)
```
This is the tricky part though the check `nxt.Type.Kind() != reflect.Interface` is just based on my observations that the type kind should be `reflect.Interface` is these cases where there is no concrete type.

And this type assertions succeeds:
```go
var v interface{}
var r io.Reader
r = bytes.NewBuffer([]byte("hello"))
v = r
_ = v.(io.Reader)
```

Let's chat offline to clear things up.